### ISSUE TITLE
Using primary constructors reports erroneous NRT warnings in some cases.

### ISSUE THREAD

**Issue by CyrusNajmabadi** [dotnet/roslyn#68616]
See build results here: https://github.com/dotnet/roslyn/pull/68299/checks?check_run_id=14264741895
For PR:  https://github.com/dotnet/roslyn/pull/68299
Specifically, commit: https://github.com/dotnet/roslyn/pull/68299/commits/4b9834b9646b027597995e45f89503012f571aef

I'll be reverting these changes to be able to move forward with that PR.  But the above commit should show the issue.

An example of the error is:

```
src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/TextChangeMerger.cs#L27
src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/TextChangeMerger.cs(27,80): error CS8602: (NETCORE_ENGINEERING_TELEMETRY=Build) Dereference of a possibly null reference.
```

The code in question is:

```c#
        private readonly IDocumentTextDifferencingService _differenceService = document.Project.Solution.Services.GetRequiredService<IDocumentTextDifferencingService>();
```

![image](https://github.com/dotnet/roslyn/assets/4564579/53fd074d-2515-4e6b-a70e-880634081af7)

I cannot determine why the compiler feels the value can be null here.  'document' is declared as a non-nullable parameter, and tthe usages of 'document' are as follows:

![image](https://github.com/dotnet/roslyn/assets/4564579/74345745-9192-488b-86e7-51e8f190f25a)

![image](https://github.com/dotnet/roslyn/assets/4564579/0cf562cf-bfa9-4c6c-aec0-435154568b25)

![image](https://github.com/dotnet/roslyn/assets/4564579/8117b1a3-b965-43ae-8e73-9caf573817be)

None of these should feed into 'document' being null at any of these points.

I'm not sure if this is a primary constructor issue, or an NRT issue.

**Comment by AlekseyTs** [comment]
@CyrusNajmabadi To many words. :-) This could be a dupe of #68384.

**Comment by CyrusNajmabadi** [comment]
Yes, seems like a strong possibility.  @mavasani could we move to a version of Roslyn with Aleksey's latest fix to try out?

**Comment by mavasani** [comment]
> Yes, seems like a strong possibility. @mavasani could we move to a version of Roslyn with Aleksey's latest fix to try out?

@CyrusNajmabadi We generally avoid moving to such latest preview releases of compiler toolsets. I am not sure if this causes any issues for community members who don't have internal preview builds of VS or .NET SDK. I think you we may just want to add a pragma suppression in IDE code for false positive compiler warnings, which can be removed later when we upgrade the compiler toolset used by the repo.

**Comment by mavasani** [comment]
I tried applying the `Use primary constructor` code fix on `CodeFixOptionsProvider` type and build succeeds without any compiler warnings. I presume this is fixed with the latest bits.
